File: aws-globalaccelerator-endpointgroup/src/test/java/software/amazon/globalaccelerator/endpointgroup/CreateHandlerTest.java
package software.amazon.globalaccelerator.endpointgroup;

import com.amazonaws.services.globalaccelerator.model.*;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.OperationStatus;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
public class CreateHandlerTest {

    @Mock
    private AmazonWebServicesClientProxy proxy;

    @Mock
    private Logger logger;

    @BeforeEach
    public void setup() {
        proxy = mock(AmazonWebServicesClientProxy.class);
        logger = mock(Logger.class);
    }


    private List<EndpointDescription> createEndpointDescription() {
        ArrayList<EndpointDescription> array = new ArrayList<EndpointDescription>();
        array.add(new EndpointDescription()
                .withClientIPPreservationEnabled(true)
                .withHealthState("HS1")
                .withEndpointId("ID1")
                .withWeight(100)
                .withHealthReason("Reason1"));
        array.add(new EndpointDescription()
                .withClientIPPreservationEnabled(false)
                .withHealthState("HS2")
                .withEndpointId("ID2")
                .withWeight(100)
                .withHealthReason("Reason2"));
        return array;
    }

    @Test
    public void handleRequest_InitialCreate_ReturnsInProgress() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                    .withStatus(AcceleratorStatus.IN_PROGRESS));
        doReturn(describeAcceleratorResult).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // return a listener
        val describeListenerResult = new DescribeListenerResult()
                .withListener(new Listener()
                    .withListenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234"));
        doReturn(describeListenerResult).when(proxy).injectCredentialsAndInvoke(any(DescribeListenerRequest.class), any());

        // the endpoint group create result
        val createEndpointGroupResult = new CreateEndpointGroupResult()
                .withEndpointGroup(new EndpointGroup()
                    .withEndpointGroupArn("ENDPOINT_GROUP_ARN")
                    .withHealthCheckPath("/MYPATH")
                    .withHealthCheckPort(200)
                    .withEndpointGroupRegion("us-west-2")
                    .withHealthCheckProtocol("HTTP")
                    .withHealthCheckIntervalSeconds(10)
                    .withThresholdCount(4)
                    .withEndpointDescriptions(createEndpointDescription()));
        doReturn(createEndpointGroupResult).when(proxy).injectCredentialsAndInvoke(any(CreateEndpointGroupRequest.class), any());

        // create the that will go to our handler
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build();

        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val handler = new CreateHandler();
        val response = handler.handleRequest(proxy, request, null, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.IN_PROGRESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
        assertThat(response.getCallbackContext()).isNotNull();
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();

        // want to confirm data was updated correctly
        assertThat(response.getResourceModel()).isNotNull();
        assertThat(response.getResourceModel().getEndpointGroupArn()).isEqualTo("ENDPOINT_GROUP_ARN");
        assertThat(response.getResourceModel().getHealthCheckPath()).isEqualTo("/MYPATH");
        assertThat(response.getResourceModel().getHealthCheckPort()).isEqualTo(200);
        assertThat(response.getResourceModel().getEndpointGroupRegion()).isEqualTo("us-west-2");
        assertThat(response.getResourceModel().getHealthCheckProtocol()).isEqualTo("HTTP");
        assertThat(response.getResourceModel().getHealthCheckIntervalSeconds()).isEqualTo(10);
        assertThat(response.getResourceModel().getThresholdCount()).isEqualTo(4);
    }

    @Test
    public void handleRequest_AwaitAcceleratorToGoInSync_ReturnsInProgress() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS));
        doReturn(describeAcceleratorResult).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // Create the model we will provide to our handler
        val endpointConfigurations = new ArrayList<EndpointConfiguration>();
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").build());
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build());
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .endpointGroupArn("ENDPOINT_GROUP")
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build();

        // call the handler
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val callbackContext = CallbackContext.builder().stabilizationRetriesRemaining(5).build();
        val handler = new CreateHandler();
        val response = handler.handleRequest(proxy, request, callbackContext, logger);

        // validate expectations
        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.IN_PROGRESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(1);
        assertThat(response.getCallbackContext()).isNotNull();
        assertThat(response.getCallbackContext().getStabilizationRetriesRemaining()).isEqualTo(4);
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();

        // confirm the model did not mutate
        assertThat(response.getResourceModel()).isEqualTo(model);
    }

    @Test
    public void handleRequest_AcceleratorIsInSync_ReturnsSuccess() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withStatus(AcceleratorStatus.DEPLOYED));
        doReturn(describeAcceleratorResult).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // Create the model we will provide to our handler
        val endpointConfigurations = new ArrayList<EndpointConfiguration>();
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").build());
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build());
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .endpointGroupArn("ENDPOINT_GROUP")
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build();

        // call the handler
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val callbackContext = CallbackContext.builder().stabilizationRetriesRemaining(5).build();
        val handler = new CreateHandler();
        val response = handler.handleRequest(proxy, request, callbackContext, logger);

        // validate expectations
        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.SUCCESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
        assertThat(response.getCallbackContext()).isNull();
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();

        // confirm the model did not mutate
        assertThat(response.getResourceModel()).isEqualTo(model);
    }

    @Test
    public void handleRequest_ThresholdTimeExceeded() {
        // Create the model we will provide to our handler
        val endpointConfigurations = new ArrayList<EndpointConfiguration>();
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").build());
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build());
        val model = ResourceModel.builder()
                .endpointGroupRegion("us-west-2")
                .listenerArn("arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234")
                .endpointGroupArn("ENDPOINT_GROUP_ARN")
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build();

        // call the handler
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();
        val callbackContext = CallbackContext.builder().stabilizationRetriesRemaining(0).build();
        val handler = new CreateHandler();

        assertThatExceptionOfType(RuntimeException.class).isThrownBy(() -> handler.handleRequest(proxy, request, callbackContext, logger))
                .withMessageMatching("Timed out waiting for endpoint group to be deployed.");
    }
}

File: aws-globalaccelerator-endpointgroup/src/test/kotlin/software/amazon/globalaccelerator/endpointgroup/CreateHandlerTest.kt
package software.amazon.globalaccelerator.endpointgroup

import com.amazonaws.services.globalaccelerator.model.EndpointGroup
import com.amazonaws.services.globalaccelerator.model.EndpointDescription
import com.amazonaws.services.globalaccelerator.model.Accelerator
import com.amazonaws.services.globalaccelerator.model.Listener
import com.amazonaws.services.globalaccelerator.model.AcceleratorStatus
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorResult
import com.amazonaws.services.globalaccelerator.model.DescribeListenerResult
import com.amazonaws.services.globalaccelerator.model.CreateEndpointGroupResult
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.OperationStatus
import software.amazon.cloudformation.proxy.ResourceHandlerRequest

import java.util.ArrayList

@ExtendWith(MockKExtension::class)
class CreateHandlerTest {
    @MockK
    lateinit var proxy: AmazonWebServicesClientProxy

    @MockK(relaxed = true)
    lateinit var logger: Logger

    @BeforeEach
    fun setup() = MockKAnnotations.init(this)

    private val endpointGroupRegion = "us-west-2"
    private val listenerArn = "arn:aws:globalaccelerator::474880776455:accelerator/abcd1234/listener/12341234"
    private val endpointGroupArn = "arn:aws:globalaccelerator::444607872184:accelerator/88127aa5-01d8-484c-80a0-349daaefce1d/listener/ee7358c2/endpoint-group/de69a4b45005"

    private fun createEndpointDescription(): List<EndpointDescription> {
        val array = ArrayList<EndpointDescription>()
        array.add(EndpointDescription()
                .withClientIPPreservationEnabled(true)
                .withHealthState("HS1")
                .withEndpointId("ID1")
                .withWeight(100)
                .withHealthReason("Reason1"))
        array.add(EndpointDescription()
                .withClientIPPreservationEnabled(false)
                .withHealthState("HS2")
                .withEndpointId("ID2")
                .withWeight(100)
                .withHealthReason("Reason2"))
        return array
    }

    @Test
    fun handleRequest_InitialCreate_ReturnsInProgress() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns describeAcceleratorResult

        // return a listener
        val describeListenerResult = DescribeListenerResult()
                .withListener(Listener()
                        .withListenerArn(listenerArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeListener>()) } returns describeListenerResult

        // the endpoint group create result
        val createEndpointGroupResult = CreateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn)
                        .withHealthCheckPath("/MYPATH")
                        .withHealthCheckPort(200)
                        .withEndpointGroupRegion(endpointGroupRegion)
                        .withHealthCheckProtocol("HTTP")
                        .withHealthCheckIntervalSeconds(10)
                        .withThresholdCount(4)
                        .withTrafficDialPercentage(100.0f)
                        .withEndpointDescriptions(createEndpointDescription())
                )
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyCreateEndpointGroup>()) } returns createEndpointGroupResult

        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .healthCheckPort(-1)
                .thresholdCount(3)
                .build()

        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val response = CreateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertEquals(0, response.callbackDelaySeconds)
        assertNotNull(response.callbackContext)
        assertNotNull(response.resourceModel)
        assertNull(response.message)
        assertNull(response.resourceModels)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
        assertFalse(response.callbackContext!!.pendingStabilization)

        assertEquals("/MYPATH", response.resourceModel.healthCheckPath)
        assertEquals(200, response.resourceModel.healthCheckPort)
        assertEquals(endpointGroupRegion, response.resourceModel.endpointGroupRegion)
        assertEquals("HTTP", response.resourceModel.healthCheckProtocol)
        assertEquals(10, response.resourceModel.healthCheckIntervalSeconds)
        assertEquals(4, response.resourceModel.thresholdCount)
        assertEquals(100.0, response.resourceModel.trafficDialPercentage)
    }

    @Test
    fun handleRequest_AwaitAcceleratorToGoInSync_ReturnsInProgress() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns describeAcceleratorResult

        // Create the model we will provide to our handler
        val endpointConfigurations = ArrayList<EndpointConfiguration>()
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").attachmentArn("ATT1").build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build())
        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100.0)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build()

        // call the handler
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(5)
        val response = CreateHandler().handleRequest(proxy, request, callbackContext, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertEquals(1, response.callbackDelaySeconds)
        assertEquals(4, response.callbackContext!!.stabilizationRetriesRemaining)
        assertNotNull(response.callbackContext)
        assertNull(response.message)
        assertNull(response.resourceModels)
        assertEquals(model, response.resourceModel)
    }

    @Test
    fun handleRequest_AcceleratorIsInSync_ReturnsSuccess() {
        // return an accelerator that is IN_PROGRESS
        val describeAcceleratorResult = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.DEPLOYED))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns describeAcceleratorResult

        // Create the model we will provide to our handler
        val endpointConfigurations = ArrayList<EndpointConfiguration>()
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").attachmentArn("ATT1").build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build())
        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100.0)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build()

        // call the handler
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(5)
        val response = CreateHandler().handleRequest(proxy, request, callbackContext, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.SUCCESS, response.status)
        assertEquals(0, response.callbackDelaySeconds)
        assertNull(response.callbackContext)
        assertNull(response.message)
        assertNull(response.resourceModels)
        assertEquals(model, response.resourceModel)

    }

    @Test
    fun handleRequest_ThresholdTimeExceeded() {
        // Create the model we will provide to our handler
        val endpointConfigurations = ArrayList<EndpointConfiguration>()
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID1").attachmentArn("ATT1").build())
        endpointConfigurations.add(EndpointConfiguration.builder().endpointId("EPID2").build())
        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .endpointGroupArn(endpointGroupArn)
                .healthCheckPort(20)
                .thresholdCount(3)
                .trafficDialPercentage(100.0)
                .healthCheckPath("/HEALTH")
                .endpointConfigurations(endpointConfigurations)
                .build()

        // call the handler
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(0)
        val handler = CreateHandler()

        val exception = assertThrows(RuntimeException::class.java) {
            handler.handleRequest(proxy, request, callbackContext, logger)
        }
        assertEquals("Timed out waiting for endpoint group to be deployed.", exception.message)
    }

    @Test
    fun handleRequest_CreateEndpointGroupWithPortOverrides_ReturnsInProgress() {
        val describeAcceleratorResult = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns describeAcceleratorResult

        val describeListenerResult = DescribeListenerResult().withListener(Listener().withListenerArn(listenerArn))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeListener>()) } returns describeListenerResult

        val returnedPortOverrides = com.amazonaws.services.globalaccelerator.model.PortOverride().withListenerPort(80).withEndpointPort(8080)
        val createEndpointGroupResult = CreateEndpointGroupResult()
                .withEndpointGroup(EndpointGroup()
                        .withEndpointGroupArn(endpointGroupArn)
                        .withHealthCheckPath("/MYPATH")
                        .withHealthCheckPort(200)
                        .withEndpointGroupRegion(endpointGroupRegion)
                        .withHealthCheckProtocol("HTTP")
                        .withHealthCheckIntervalSeconds(10)
                        .withThresholdCount(4)
                        .withTrafficDialPercentage(100.0f)
                        .withEndpointDescriptions(createEndpointDescription())
                        .withPortOverrides(returnedPortOverrides)
                )
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyCreateEndpointGroup>()) } returns createEndpointGroupResult

        val portOverrides = listOf(PortOverride(80, 8080))
        val model = ResourceModel.builder()
                .endpointGroupRegion(endpointGroupRegion)
                .listenerArn(listenerArn)
                .healthCheckPort(-1)
                .thresholdCount(3)
                .portOverrides(portOverrides)
                .build()

        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val response = CreateHandler().handleRequest(proxy, request, null, logger)

        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertEquals(0, response.callbackDelaySeconds)
        assertNotNull(response.callbackContext)
        assertNotNull(response.resourceModel)
        assertNull(response.message)
        assertNull(response.resourceModels)
        assertEquals(endpointGroupArn, response.resourceModel.endpointGroupArn)
        assertFalse(response.callbackContext!!.pendingStabilization)

        assertEquals("/MYPATH", response.resourceModel.healthCheckPath)
        assertEquals(200, response.resourceModel.healthCheckPort)
        assertEquals(endpointGroupRegion, response.resourceModel.endpointGroupRegion)
        assertEquals("HTTP", response.resourceModel.healthCheckProtocol)
        assertEquals(10, response.resourceModel.healthCheckIntervalSeconds)
        assertEquals(4, response.resourceModel.thresholdCount)
        assertEquals(100.0, response.resourceModel.trafficDialPercentage)
        assertEquals(1, response.resourceModel.portOverrides.size)
        assertEquals(returnedPortOverrides.listenerPort, response.resourceModel.portOverrides[0].listenerPort)
        assertEquals(returnedPortOverrides.endpointPort, response.resourceModel.portOverrides[0].endpointPort)
    }
}
